File: broker.go/internal/bridge/mqttclient/manager.go
package mqttclient

import (
	"context"
	"encoding/json"
	"log/slog"
	"sync"

	"monstermq.io/edge/internal/stores"
)

// Manager loads MQTT-bridge device configs from the DeviceConfigStore and runs
// one Connector per enabled device.
type Manager struct {
	store     stores.DeviceConfigStore
	publisher LocalPublisher
	subBus    LocalSubscriber
	logger    *slog.Logger
	nodeID    string

	mu         sync.Mutex
	connectors map[string]*Connector
	lastConfig map[string]string // device name → raw JSON last deployed
}

func NewManager(store stores.DeviceConfigStore, publisher LocalPublisher, subBus LocalSubscriber, nodeID string, logger *slog.Logger) *Manager {
	return &Manager{
		store: store, publisher: publisher, subBus: subBus, logger: logger, nodeID: nodeID,
		connectors: map[string]*Connector{},
		lastConfig: map[string]string{},
	}
}

func (m *Manager) Start(ctx context.Context) error {
	devices, err := m.store.GetEnabledByNode(ctx, m.nodeID)
	if err != nil {
		return err
	}
	for _, d := range devices {
		if d.Type != "" && d.Type != "MQTT_CLIENT" {
			continue
		}
		var cfg Config
		if err := json.Unmarshal([]byte(d.Config), &cfg); err != nil {
			m.logger.Warn("bridge config parse failed", "name", d.Name, "err", err)
			continue
		}
		if cfg.ClientID == "" {
			cfg.ClientID = "edge-" + m.nodeID + "-" + d.Name
		}
		c := NewConnector(d.Name, cfg, m.publisher, m.subBus, m.logger)
		if err := c.Start(ctx); err != nil {
			m.logger.Warn("bridge start failed", "name", d.Name, "err", err)
			continue
		}
		m.mu.Lock()
		m.connectors[d.Name] = c
		m.lastConfig[d.Name] = d.Config
		m.mu.Unlock()
	}
	return nil
}

func (m *Manager) Stop() {
	m.mu.Lock()
	defer m.mu.Unlock()
	for _, c := range m.connectors {
		c.Stop()
	}
	m.connectors = map[string]*Connector{}
}

// Reload reconciles the live connector set against the persisted device
// configs: starts new ones, restarts changed ones, stops removed/disabled
// ones. Call after every GraphQL mutation that touches an MQTT client.
func (m *Manager) Reload(ctx context.Context) error {
	devices, err := m.store.GetEnabledByNode(ctx, m.nodeID)
	if err != nil {
		return err
	}
	wanted := map[string]Config{}
	wantedRaw := map[string]string{}
	for _, d := range devices {
		if d.Type != "" && d.Type != "MQTT_CLIENT" {
			continue
		}
		var cfg Config
		if err := json.Unmarshal([]byte(d.Config), &cfg); err != nil {
			m.logger.Warn("bridge config parse failed", "name", d.Name, "err", err)
			continue
		}
		if cfg.ClientID == "" {
			cfg.ClientID = "edge-" + m.nodeID + "-" + d.Name
		}
		wanted[d.Name] = cfg
		wantedRaw[d.Name] = d.Config
	}

	m.mu.Lock()
	current := m.connectors
	currentRaw := m.lastConfig
	m.mu.Unlock()

	// Stop bridges that are no longer wanted, or whose stored config changed.
	for name, c := range current {
		if newRaw, keep := wantedRaw[name]; !keep || newRaw != currentRaw[name] {
			c.Stop()
			m.mu.Lock()
			delete(m.connectors, name)
			delete(m.lastConfig, name)
			m.mu.Unlock()
		}
	}

	// Start bridges that should be running but aren't.
	for name, cfg := range wanted {
		m.mu.Lock()
		_, running := m.connectors[name]
		m.mu.Unlock()
		if running {
			continue
		}
		c := NewConnector(name, cfg, m.publisher, m.subBus, m.logger)
		if err := c.Start(ctx); err != nil {
			m.logger.Warn("bridge start failed", "name", name, "err", err)
			continue
		}
		m.mu.Lock()
		m.connectors[name] = c
		m.lastConfig[name] = wantedRaw[name]
		m.mu.Unlock()
	}
	return nil
}

func (m *Manager) Active() []string {
	m.mu.Lock()
	defer m.mu.Unlock()
	out := make([]string, 0, len(m.connectors))
	for name := range m.connectors {
		out = append(out, name)
	}
	return out
}

File: broker.rs/src/bridge/mqttclient/manager.rs
use super::connector::{Config as BridgeConfig, Connector, LocalPublishFn};
use crate::pubsub::Bus as PubsubBus;
use crate::stores::traits::DeviceConfigStore;
use parking_lot::Mutex;
use std::collections::HashMap;
use std::sync::Arc;

/// Loads MQTT-bridge device configs from the DeviceConfigStore and runs one
/// Connector per enabled device. Mirrors `internal/bridge/mqttclient/manager.go`.
pub struct Manager {
    store: Arc<dyn DeviceConfigStore>,
    publisher: LocalPublishFn,
    pubsub: Arc<PubsubBus>,
    pub node_id: String,
    inner: Mutex<Inner>,
}

struct Inner {
    connectors: HashMap<String, Arc<Connector>>,
    last_config: HashMap<String, String>,
}

impl Manager {
    pub fn new(store: Arc<dyn DeviceConfigStore>, publisher: LocalPublishFn, pubsub: Arc<PubsubBus>, node_id: String) -> Arc<Self> {
        Arc::new(Self {
            store, publisher, pubsub, node_id,
            inner: Mutex::new(Inner { connectors: HashMap::new(), last_config: HashMap::new() }),
        })
    }

    pub async fn start(self: Arc<Self>) -> anyhow::Result<()> {
        let devices = self.store.get_enabled_by_node(&self.node_id).await?;
        for d in devices {
            if !d.r#type.is_empty() && d.r#type != "MQTT_CLIENT" { continue; }
            let cfg: BridgeConfig = match serde_json::from_str(&d.config) {
                Ok(c) => c,
                Err(e) => { tracing::warn!(name = d.name, err = ?e, "bridge config parse failed"); continue; }
            };
            let mut cfg = cfg;
            if cfg.client_id.is_empty() {
                cfg.client_id = format!("edge-{}-{}", self.node_id, d.name);
            }
            let c = Connector::new(d.name.clone(), cfg, self.publisher.clone(), self.pubsub.clone());
            if let Err(e) = c.clone().start().await {
                tracing::warn!(name = d.name, err = ?e, "bridge start failed");
                continue;
            }
            let mut g = self.inner.lock();
            g.connectors.insert(d.name.clone(), c);
            g.last_config.insert(d.name.clone(), d.config.clone());
        }
        Ok(())
    }

    pub async fn stop(&self) {
        let conns: Vec<Arc<Connector>> = self.inner.lock().connectors.values().cloned().collect();
        for c in conns { c.stop().await; }
        self.inner.lock().connectors.clear();
    }

    /// Reconciles the live connector set against the persisted state.
    /// Call after every GraphQL mutation that touches an MQTT client.
    pub async fn reload(self: Arc<Self>) -> anyhow::Result<()> {
        let devices = self.store.get_enabled_by_node(&self.node_id).await?;
        let mut wanted: HashMap<String, BridgeConfig> = HashMap::new();
        let mut wanted_raw: HashMap<String, String> = HashMap::new();
        for d in devices {
            if !d.r#type.is_empty() && d.r#type != "MQTT_CLIENT" { continue; }
            let cfg: BridgeConfig = match serde_json::from_str(&d.config) {
                Ok(c) => c,
                Err(e) => { tracing::warn!(name = d.name, err = ?e, "bridge config parse failed"); continue; }
            };
            let mut cfg = cfg;
            if cfg.client_id.is_empty() {
                cfg.client_id = format!("edge-{}-{}", self.node_id, d.name);
            }
            wanted.insert(d.name.clone(), cfg);
            wanted_raw.insert(d.name, d.config);
        }
        // Stop bridges that are no longer wanted, or whose config changed.
        let to_stop: Vec<(String, Arc<Connector>)> = {
            let g = self.inner.lock();
            g.connectors.iter().filter_map(|(name, c)| {
                let stop = match wanted_raw.get(name) {
                    Some(raw) => g.last_config.get(name).map(|prev| raw != prev).unwrap_or(true),
                    None => true,
                };
                if stop { Some((name.clone(), c.clone())) } else { None }
            }).collect()
        };
        for (name, c) in to_stop {
            c.stop().await;
            let mut g = self.inner.lock();
            g.connectors.remove(&name);
            g.last_config.remove(&name);
        }
        // Start bridges that should be running but aren't.
        for (name, cfg) in wanted {
            if self.inner.lock().connectors.contains_key(&name) { continue; }
            let c = Connector::new(name.clone(), cfg, self.publisher.clone(), self.pubsub.clone());
            if let Err(e) = c.clone().start().await {
                tracing::warn!(name, err = ?e, "bridge start failed");
                continue;
            }
            let mut g = self.inner.lock();
            g.connectors.insert(name.clone(), c);
            if let Some(raw) = wanted_raw.get(&name) {
                g.last_config.insert(name, raw.clone());
            }
        }
        Ok(())
    }

    pub fn active(&self) -> Vec<String> {
        self.inner.lock().connectors.keys().cloned().collect()
    }
}
